src/builtins.rs

```diff
@@ -257,9 +258,142 @@ lazy_static! {
                 }
             }
         ),
+        (
+            "LOWER_BOUND",
+            BuiltIn {
+                decl: "FUNCTION LOWER_BOUND<U: ANY> : DINT
+                VAR_INPUT
+                    arr : U;
+                    dim : DINT;
+                END_VAR
+                END_FUNCTION",
+                annotation: None,
+                generic_name_resolver: no_generic_name_resolver,
+                code : |generator, params, location| {
+                    generate_variable_length_array_bound_function(generator, params, location, true)
+                }
+            }
+        ),
+        (
+            "UPPER_BOUND",
+            BuiltIn {
+                decl: "FUNCTION UPPER_BOUND<U: ANY> : DINT
+                VAR_INPUT
+                    arr : U;
+                    dim : DINT;
+                END_VAR
+                END_FUNCTION",
+                annotation: None,
+                generic_name_resolver: no_generic_name_resolver,
+                code : |generator, params, location| {
+                    generate_variable_length_array_bound_function(generator, params, location, false)
+                }
+            }
+        ),
     ]);
 }
 
+/// Generates the code for the LOWER- AND UPPER_BOUND built-in functions, returning an error if the function
+/// arguments are incorrect.
+fn generate_variable_length_array_bound_function<'ink>(
+    generator: &ExpressionCodeGenerator<'ink, '_>,
+    params: &[&AstStatement],
+    location: SourceRange,
+    is_lower: bool,
+) -> Result<ExpressionValue<'ink>, Diagnostic> {
+    let [AstStatement::Reference { .. }, AstStatement::LiteralInteger {..} | AstStatement::Reference {..}] = params else {
+        return Err(
+            Diagnostic::codegen_error(
+                &format!("Invalid {} function signature", if is_lower { "LOWER_BOUND" } else { "UPPER_BOUND" }),
+                location
+            )
+        )
+    };
```

Comment:
This could also be solved by adding a diagnostician to the built-ins. It can be validated, but currently it is not possible to report this (outside of verbosely checking during codegen instead of using the annotator)

Reply:
so we would abort codegen here? the cleanest solution would be if we can do this during the validation phase :thinking:

Reply:
Added a validator to the builtins. It's now possible to have custom validation for all builtin functions. I have basically mirrored the annotation/resolver implementation, but chose to have this validation in addition to any other validation that takes place instead of branching.
I also removed the returned Result<(), Diagnostic> from the annotation, since the resolver which is supposed to handle this is not equipped to do so (so far we just blindly unwrapped anyway).